src/numeric/decision_rules.jl

```diff
@@ -17,190 +51,169 @@ function filter_mcoeffs(a::Array{Float64,1},b::Array{Float64,1},n::Array{Int64,1
+    return (DecisionRule{UnstructuredGrid, CartesianGrid})(grid_exo, grid_endo, n_x, coeffs)
 end
-
-function DecisionRule(dp::DiscretizedProcess, grid_endo::CartesianGrid, n_x::Int)
-    # this assumes the discretized process
-    # was discretized on a cartesian grid
-    a_exo = minimum(dp.nodes, 1)[:]
-    b_exo = maximum(dp.nodes, 1)[:]
-    o_exo = Int64[length(unique(dp.nodes[:, i])) for i=1:size(dp.nodes, 2)]
-    grid_exo = CartesianGrid(a_exo, b_exo, o_exo)
-    grid = CartesianGrid(
-            cat(1, grid_exo.min, grid_endo.min),
-            cat(1, grid_exo.max, grid_endo.max),
-            cat(1, grid_exo.n, grid_endo.n)
-        )
-    a = grid.min
-    b = grid.max
-    orders = grid.n
-    # n_ms = size(dp.values, 1)
-    coeffs = [zeros(n_x, (orders+2)...)]
-    return CPDecisionRule(dp, grid_exo, grid_endo, grid, n_x, coeffs)
-end
-
-function DecisionRule(dp::DiscretizedProcess, grid::CartesianGrid, values::Array{Array{Float64,2}})
+#
+function DecisionRule(grid_exo::UnstructuredGrid, grid_endo::CartesianGrid, values::Array{Array{Float64,2}})
     n_x = size(values[1], 2)
-    dr = DecisionRule(dp, grid, n_x)
+    dr = DecisionRule(grid_exo, grid_endo, n_x)
     set_values!(dr, values)
     return dr
 end
 
-function (dr::CPDecisionRule)(z::Array{Float64,2})
-    a = dr.grid.min
-    b = dr.grid.max
-    n = dr.grid.n
-    cc = dr.coefficients[1]
-    res = splines.eval_UC_multi_spline(a, b, n, cc, z)'
-    return res
-end
 
-(dr::CPDecisionRule)(z::Vector) = dr(z')[:]
+function set_values!(dr::AbstractDecisionRule{UnstructuredGrid, CartesianGrid}, values::Array{Matrix{Float64},1})
+    a = dr.grid_endo.min
+    b = dr.grid_endo.max
+    orders = dr.grid_endo.n
+    dr.coefficients = [filter_mcoeffs(a, b, orders, vals) for vals in values]
+end
 
 
-function (dr::CPDecisionRule)(i::Int64, x::Array{Float64,2})
-    a = dr.grid.min
-    b = dr.grid.max
-    n = dr.grid.n
-    cc = dr.coefficients[1]
-    N = size(x, 1)
-    e = dr.process.nodes[i, :]
-    z = cat(2, repmat(e', N, 1), x)
+function evaluate(dr::AbstractDecisionRule{UnstructuredGrid, CartesianGrid}, i::Int64, z::Matrix{Float64})
+    a = dr.grid_endo.min
+    b = dr.grid_endo.max
+    n = dr.grid_endo.n
+    cc = dr.coefficients[i]
     res = splines.eval_UC_multi_spline(a, b, n, cc, z)'
     return res
 end
 
+(dr::DecisionRule{UnstructuredGrid, CartesianGrid})(i::Int64,y::Matrix{Float64}) = evaluate(dr,i,y)
+(dr::DecisionRule{UnstructuredGrid, CartesianGrid})(i::Int64,y::Vector{Float64}) = dr(i,y')[:]
 
 
-function (dr::CPDecisionRule)(i::Int64, j::Int64, x::Array{Float64,2})
-    a = dr.grid.min
-    b = dr.grid.max
-    n = dr.grid.n
-    cc = dr.coefficients[1]
-    N = size(x, 1)
-    e = dr.process.integration_nodes[i][j, :]
-    z = cat(2, repmat(e', N, 1), x)
-    res = splines.eval_UC_multi_spline(a, b, n, cc, z)'
-    return res
+
+
+#####
+##### Cached Decision Rules (do we really need them ?)
```

Comment:
I don't have a great sense of how they are used (yet), so I can't think of any better names